src/main/kotlin/org/radarbase/push/integration/garmin/service/BackfillService.kt

```diff
@@ -11,9 +28,86 @@ import javax.ws.rs.core.Context
  * service to create POST requests for historic data to our server.
  */
 class BackfillService(
-    @Context config: Config,
-    @Named(GARMIN_QUALIFIER) userRepository: GarminUserRepository
-) {
-    // TODO: discuss about this since we need some way to store the offset for historic requests
-    //  made for each user
+    @Context private val config: Config,
+    @Named(GARMIN_QUALIFIER) private val userRepository: GarminUserRepository
+) : ApplicationEventListener {
+
+    private val redisHolder =
+        RedisHolder(JedisPool(config.pushIntegration.garmin.backfill.redis.uri))
+    private val executorService = Executors.newSingleThreadScheduledExecutor()
+    private val requestExecutorService = Executors.newFixedThreadPool(
+        config.pushIntegration.garmin.backfill.maxThreads
+    )
+    private val requestGenerator = GarminRequestGenerator(config, userRepository)
+    private val remoteLockManager = RedisRemoteLockManager(
+        redisHolder,
+        config.pushIntegration.garmin.backfill.redis.lockPrefix
+    )
+    private val httpClient = OkHttpClient()
+    private val requestsPerUserPerIteration: Int
+        get() = 40
+
+    private val futures: MutableList<Future<*>> = mutableListOf()
+
+    override fun onEvent(event: ApplicationEvent?) {
+        when (event?.type) {
+            INITIALIZATION_FINISHED -> start()
+            DESTROY_FINISHED -> stop()
+            else -> logger.info("Application event received: ${event?.type}")
+        }
+    }
+
+    override fun onRequest(requestEvent: RequestEvent?): RequestEventListener? = null
+
+    private fun start() {
+        logger.info("Application Initialisation completed. Starting Backfill service...")
+
+        executorService.scheduleAtFixedRate(::iterateUsers, 1, 5, TimeUnit.MINUTES)
+    }
+
+    private fun stop() {
+        logger.info("Application Destroy completed. Stopping Backfill service...")
+        try {
+            requestExecutorService.awaitTermination(30, TimeUnit.SECONDS)
+            executorService.awaitTermination(30, TimeUnit.SECONDS)
+        } catch (e: InterruptedException) {
+            logger.error("Failed to complete execution: interrupted")
+        }
+    }
+
+    private fun iterateUsers() {
```

Comment:
Please catch all exceptions from this method and log them, since otherwise the scheduler thread could be affected.

```kotlin
try {
    // make request
} catch (ex: Throwable) {
    // log something
}
```